src/tyr/serializers.cc

```diff
@@ -209,6 +224,60 @@ json::ArrayPtr waypoints(const valhalla::Trip& trip) {
   return waypoints;
 }
 
+/*
+ * This function takes any waypoints (excluding origin and destination) and walks the shape
+ * for a match.  When a match is found, we store the geometry index.
+ * We use the valhalla midgard `equal` function to check for equality based on precision.
+ */
+json::ArrayPtr via_waypoints(valhalla::Options options, const std::vector<PointLL>& shape) {
+  // Create a vector of indexes based on the number of locations.
+  auto locs = *options.mutable_locations();
+  uint32_t i = 0;
+  std::vector<uint32_t> location_indexes;
+  for (const auto& loc : locs) {
+    location_indexes.push_back(i++);
+  }
+  auto via_waypoints = json::array({});
+  for (const auto& location_index : location_indexes) {
+    if (location_index == 0 || location_index == (locs.size() - 1)) {
+      continue;
+    }
+
+    // Only create via_waypoints object if the locations are via or through types
+    if (locs.Get(location_index).type() == valhalla::Location::kVia ||
+        locs.Get(location_index).type() == valhalla::Location::kThrough) {
+      locs.Mutable(location_index)->set_waypoint_index(location_index);
+
+      double distance_from_leg_start = 0;
+      uint32_t geometry_idx = 0;
+      for (const auto& ll : shape) {
```

Comment:
You could also adjust this second loop to not re-calculate the done segments over and over, and instead continue where the last waypoint was. It could matter in the case where there are a large number of via waypoints to not have to recalculate number of coordinates times number of waypoints.